libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/transforms/ClasspathEntrySnapshotTransform.kt

```diff
@@ -6,26 +6,97 @@
+abstract class ClasspathEntrySnapshotTransform : TransformAction<ClasspathEntrySnapshotTransform.Parameters> {
+
+    abstract class Parameters : TransformParameters {
+        @get:Internal
+        abstract val gradleUserHomeDir: DirectoryProperty
+
+        @get:Internal
+        abstract val buildMetricsReporterService: Property<BuildMetricsReporterService>
+    }
 
     @get:Classpath
     @get:InputArtifact
     abstract val inputArtifact: Provider<FileSystemLocation>
 
     override fun transform(outputs: TransformOutputs) {
-        val classpathEntry = inputArtifact.get().asFile
-        val snapshotFile = outputs.file(classpathEntry.name.replace('.', '_') + "-snapshot.bin")
+        val classpathEntryInputDirOrJar = inputArtifact.get().asFile
+        val snapshotOutputFile = outputs.file(classpathEntryInputDirOrJar.name.replace('.', '_') + "-snapshot.bin")
+
+        val granularity = getClassSnapshotGranularity(classpathEntryInputDirOrJar, parameters.gradleUserHomeDir.get().asFile)
+
+        val buildMetricsReporterService = parameters.buildMetricsReporterService.orNull
+        val metricsReporter = buildMetricsReporterService?.let { BuildMetricsReporterImpl() } ?: DoNothingBuildMetricsReporter
+
+        val startTimeMs = System.currentTimeMillis()
+        var failureMessage: String? = null
+        try {
+            doTransform(classpathEntryInputDirOrJar, snapshotOutputFile, granularity, metricsReporter)
+        } catch (e: Throwable) {
+            failureMessage = e.message
+            throw e
+        } finally {
+            buildMetricsReporterService?.addTransformMetrics(
+                transformPath = "${ClasspathEntrySnapshotTransform::class.simpleName} for ${classpathEntryInputDirOrJar.path}",
+                transformClass = ClasspathEntrySnapshotTransform::class.java,
+                isKotlinTransform = true,
+                startTimeMs = startTimeMs,
+                totalTimeMs = System.currentTimeMillis() - startTimeMs,
+                buildMetrics = metricsReporter.getMetrics(),
+                failureMessage = failureMessage
+            )
+        }
+    }
+
+    /**
+     * Determines the [ClassSnapshotGranularity] when taking a snapshot of the given [classpathEntryDirOrJar].
+     *
+     * As mentioned in [ClassSnapshotGranularity]'s kdoc, we will take [CLASS_LEVEL] snapshots for classes that are infrequently changed
+     * (e.g., external libraries which are typically stored/transformed inside the Gradle user home, or a few hard-coded cases), and take
+     * [CLASS_MEMBER_LEVEL] snapshots for the others.
+     */
+    private fun getClassSnapshotGranularity(classpathEntryDirOrJar: File, gradleUserHomeDir: File): ClassSnapshotGranularity {
+        return if (
+            classpathEntryDirOrJar.startsWith(gradleUserHomeDir) ||
+            classpathEntryDirOrJar.name == "android.jar"
+        ) CLASS_LEVEL
+        else CLASS_MEMBER_LEVEL
```

Comment:
I wonder if it is possible to have in the future better approach here :thinking: 

 For example, now it is ignores jars included via `implementation files("path/to/dep.jar")`.

Reply:
Yeah ideally if we have an [ArtifactCollection](https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/ArtifactCollection.html) for the classpath, we will be able to tell if a classpath entry ([ComponentIdentifier](https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/component/ComponentIdentifier.html)) is a [LibraryBinaryIdentifier](https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/component/LibraryBinaryIdentifier.html), [ModuleComponentIdentifier](https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/component/ModuleComponentIdentifier.html), or [ProjectComponentIdentifier](https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/component/ProjectComponentIdentifier.html), and handle them accordingly.

I guess we have to rely on heuristics for now until Gradle provides some new features that can help with this.